### ISSUE TITLE
Recent docker.io rate-limiting results in failing mounts

### ISSUE THREAD

**Issue by bretm** [democratic-csi/democratic-csi#90]
I'm running an OKD4 cluster, and had two updates pending.  I also use democratic-csi to provide both iscsi & nfs persistent volumes.  When I tried to do both cluster upgrades in the same day, I got some rate-limiting errors per:

  https://www.docker.com/increase-rate-limits

Ultimately, there seemed to be a cascade failure where when democratic-csi operator became unhappy, that in turn made it difficult for the image-registry cluster operator to function, leading to general unhappiness and upgrade breakage.

As this is a homelab cluster, I didn't worry about it much this weekend.  When I came back to look at it today, it seems like the rate-limiting has been overcome, but the image-registry is still offline.  I see events like this within the openshift-image-registry project:

>   MountVolume.MountDevice failed for volume "pvc-859f35da-c5b3-4f37-be64-c939669565aa" : kubernetes.io/csi: attacher.MountDevice failed to create newCsiDriverClient: driver name org.democratic-csi.nfs not found in the list of registered CSI drivers

I'm not sure what percussive maintenance will help getting rid of this error...

**Comment by travisghansen** [comment]
Hmm that’s pretty odd. I’m pretty sure that should be coming from the socket file in the kubelet dir. Can you see if on the host it has a valid socket file in the kubelet plugins dir?

**Comment by bretm** [comment]
Issue disappeared before I could get time to get in and do a deeper debugging session :/

**Comment by travisghansen** [comment]
I have it on my list to migrate to something other than docker hub...just not sure where yet.

**Comment by bretm** [comment]
https://www.redhat.com/sysadmin/container-quayio ; though, in full disclosure, I'm biased ;)

**Comment by travisghansen** [comment]
Does quay support pull stats and have an exporter by chance to get those stats?

**Comment by bretm** [comment]
Happening again on my upgrade today to 4.7.0-0.okd-2021-05-22-050008:
![image](https://user-images.githubusercontent.com/11554/119268405-ca796380-bbc0-11eb-8900-ec91018e369e.png)

I'm still learning how to debug clusters more effectively, so if there are specific diagnostics I can capture, let me know.

**Comment by travisghansen** [comment]
So what happens is the node pod creates a socket (the csi grpc socket) on the host at a specific path in the kubelet dir. If that file is not there (ie: pod hasn’t started, kubelet in a weird path, etc) then kubelet won’t detect the ‘plugin’ and is likely the result of the error.

**Comment by bretm** [comment]
> 
> 
> Does quay support pull stats and have an exporter by chance to get those stats?

Not sure; it does provide usage logs, but haven't seen an export function.  The full API is documented here:  https://docs.quay.io/api/swagger/#

**Comment by bretm** [comment]
> 
> 
> So what happens is the node pod creates a socket (the csi grpc socket) on the host at a specific path in the kubelet dir. If that file is not there (ie: pod hasn’t started, kubelet in a weird path, etc) then kubelet won’t detect the ‘plugin’ and is likely the result of the error.

So, would the various rolls of Cluster Operators during an OpenShift upgrade sweep the leg on the node pods?

**Comment by travisghansen** [comment]
Looking for something of this nature: https://github.com/infinityworks/docker-hub-exporter

**Comment by travisghansen** [comment]
Well, is the node pod up and running according to k8s? If so then something seems off somewhere as the socket should be present etc.

When I get back to my desk I’ll send you an example path where the socket should be on the host.

**Comment by bretm** [comment]
Here's what I'm seeing in the "democratic-csi" namespace:

![image](https://user-images.githubusercontent.com/11554/119268988-6a37f100-bbc3-11eb-8169-4772799cff28.png)

> [root@okd4-services ~]# oc get events
> LAST SEEN   TYPE      REASON             OBJECT                                                        MESSAGE
> 50m         Normal    LeaderElection     lease/external-resizer-org-democratic-csi-iscsi               zfs-iscsi-democratic-csi-controller-64fc47b6dd-8t9sl became leader
> 50m         Normal    LeaderElection     lease/external-resizer-org-democratic-csi-nfs                 freenas-nfs-democratic-csi-controller-697cb8cd9b-886h7 became leader
> 50m         Normal    LeaderElection     lease/external-snapshotter-leader-org-democratic-csi-iscsi    zfs-iscsi-democratic-csi-controller-64fc47b6dd-8t9sl became leader
> 50m         Normal    LeaderElection     lease/external-snapshotter-leader-org-democratic-csi-nfs      freenas-nfs-democratic-csi-controller-697cb8cd9b-886h7 became leader
> 50m         Normal    Scheduled          pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-886h7    Successfully assigned democratic-csi/freenas-nfs-democratic-csi-controller-697cb8cd9b-886h7 to node-b1.okd4.rdu.bretm.net
> 50m         Normal    AddedInterface     pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-886h7    Add eth0 [10.128.3.151/23]
> 50m         Normal    Pulled             pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-886h7    Container image "k8s.gcr.io/sig-storage/csi-provisioner:v2.1.0" already present on machine
> 50m         Normal    Created            pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-886h7    Created container external-provisioner
> 50m         Normal    Started            pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-886h7    Started container external-provisioner
> 50m         Normal    Pulled             pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-886h7    Container image "k8s.gcr.io/sig-storage/csi-resizer:v1.1.0" already present on machine
> 50m         Normal    Created            pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-886h7    Created container external-resizer
> 50m         Normal    Started            pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-886h7    Started container external-resizer
> 50m         Normal    Pulled             pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-886h7    Container image "k8s.gcr.io/sig-storage/csi-snapshotter:v3.0.3" already present on machine
> 50m         Normal    Created            pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-886h7    Created container external-snapshotter
> 50m         Normal    Started            pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-886h7    Started container external-snapshotter
> 50m         Normal    Pulling            pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-886h7    Pulling image "democraticcsi/democratic-csi:latest"
> 50m         Normal    Pulled             pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-886h7    Successfully pulled image "democraticcsi/democratic-csi:latest" in 794.25342ms
> 50m         Normal    Created            pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-886h7    Created container csi-driver
> 50m         Normal    Started            pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-886h7    Started container csi-driver
> 46m         Normal    Killing            pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-886h7    Stopping container external-provisioner
> 46m         Normal    Killing            pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-886h7    Stopping container csi-driver
> 46m         Normal    Killing            pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-886h7    Stopping container external-snapshotter
> 46m         Normal    Killing            pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-886h7    Stopping container external-resizer
> 46m         Normal    Scheduled          pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-fgnqj    Successfully assigned democratic-csi/freenas-nfs-democratic-csi-controller-697cb8cd9b-fgnqj to node-a1.okd4.rdu.bretm.net
> 46m         Normal    AddedInterface     pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-fgnqj    Add eth0 [10.131.0.11/23]
> 46m         Normal    Pulled             pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-fgnqj    Container image "k8s.gcr.io/sig-storage/csi-provisioner:v2.1.0" already present on machine
> 46m         Normal    Created            pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-fgnqj    Created container external-provisioner
> 46m         Normal    Started            pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-fgnqj    Started container external-provisioner
> 46m         Normal    Pulled             pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-fgnqj    Container image "k8s.gcr.io/sig-storage/csi-resizer:v1.1.0" already present on machine
> 46m         Normal    Created            pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-fgnqj    Created container external-resizer
> 46m         Normal    Started            pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-fgnqj    Started container external-resizer
> 46m         Normal    Pulled             pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-fgnqj    Container image "k8s.gcr.io/sig-storage/csi-snapshotter:v3.0.3" already present on machine
> 46m         Normal    Created            pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-fgnqj    Created container external-snapshotter
> 46m         Normal    Started            pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-fgnqj    Started container external-snapshotter
> 44m         Normal    Pulling            pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-fgnqj    Pulling image "democraticcsi/democratic-csi:latest"
> 42m         Warning   Failed             pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-fgnqj    Failed to pull image "democraticcsi/democratic-csi:latest": rpc error: code = Unknown desc = Error reading manifest latest in docker.io/democraticcsi/democratic-csi: toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit
> 16m         Warning   Failed             pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-fgnqj    Error: ErrImagePull
> 42m         Normal    BackOff            pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-fgnqj    Back-off pulling image "democraticcsi/democratic-csi:latest"
> 99s         Warning   Failed             pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-fgnqj    Error: ImagePullBackOff
> 50m         Normal    Killing            pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-w8hzg    Stopping container external-provisioner
> 50m         Normal    Killing            pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-w8hzg    Stopping container csi-driver
> 50m         Normal    Killing            pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-w8hzg    Stopping container external-resizer
> 50m         Normal    Killing            pod/freenas-nfs-democratic-csi-controller-697cb8cd9b-w8hzg    Stopping container external-snapshotter
> 50m         Normal    SuccessfulCreate   replicaset/freenas-nfs-democratic-csi-controller-697cb8cd9b   Created pod: freenas-nfs-democratic-csi-controller-697cb8cd9b-886h7
> 46m         Normal    SuccessfulCreate   replicaset/freenas-nfs-democratic-csi-controller-697cb8cd9b   Created pod: freenas-nfs-democratic-csi-controller-697cb8cd9b-fgnqj
> 42m         Warning   NodeNotReady       pod/freenas-nfs-democratic-csi-node-qh78f                     Node is not ready
> 38m         Normal    Pulling            pod/freenas-nfs-democratic-csi-node-qh78f                     Pulling image "democraticcsi/democratic-csi:latest"
> 40m         Warning   Failed             pod/freenas-nfs-democratic-csi-node-qh78f                     Failed to pull image "democraticcsi/democratic-csi:latest": rpc error: code = Unknown desc = Error loading manifest for target platform: Error reading manifest sha256:99b0fc04d54f4a2cb02bed4fc66d8b82fdd2c081d3f9a617a6a527ca58d2e35f in docker.io/democraticcsi/democratic-csi: toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit
> 37m         Warning   Failed             pod/freenas-nfs-democratic-csi-node-qh78f                     Error: ErrImagePull
> 40m         Normal    Pulled             pod/freenas-nfs-democratic-csi-node-qh78f                     Container image "k8s.gcr.io/sig-storage/csi-node-driver-registrar:v2.1.0" already present on machine
> 40m         Normal    Created            pod/freenas-nfs-democratic-csi-node-qh78f                     Created container driver-registrar
> 40m         Normal    Started            pod/freenas-nfs-democratic-csi-node-qh78f                     Started container driver-registrar
> 40m         Normal    Pulled             pod/freenas-nfs-democratic-csi-node-qh78f                     Container image "docker.io/busybox:1.32.0" already present on machine
> 40m         Normal    Created            pod/freenas-nfs-democratic-csi-node-qh78f                     Created container cleanup
> 40m         Normal    Started            pod/freenas-nfs-democratic-csi-node-qh78f                     Started container cleanup
> 37m         Normal    BackOff            pod/freenas-nfs-democratic-csi-node-qh78f                     Back-off pulling image "democraticcsi/democratic-csi:latest"
> 16m         Warning   Failed             pod/freenas-nfs-democratic-csi-node-qh78f                     Error: ImagePullBackOff
> 35m         Warning   Failed             pod/freenas-nfs-democratic-csi-node-qh78f                     Failed to pull image "democraticcsi/democratic-csi:latest": rpc error: code = Unknown desc = Error reading manifest latest in docker.io/democraticcsi/democratic-csi: toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit
> 48m         Warning   NodeNotReady       pod/freenas-nfs-democratic-csi-node-rppvz                     Node is not ready
> 43m         Normal    Pulling            pod/freenas-nfs-democratic-csi-node-rppvz                     Pulling image "democraticcsi/democratic-csi:latest"
> 45m         Warning   Failed             pod/freenas-nfs-democratic-csi-node-rppvz                     Failed to pull image "democraticcsi/democratic-csi:latest": rpc error: code = Unknown desc = Error loading manifest for target platform: Error reading manifest sha256:99b0fc04d54f4a2cb02bed4fc66d8b82fdd2c081d3f9a617a6a527ca58d2e35f in docker.io/democraticcsi/democratic-csi: toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit
> 15m         Warning   Failed             pod/freenas-nfs-democratic-csi-node-rppvz                     Error: ErrImagePull
> 45m         Normal    Pulled             pod/freenas-nfs-democratic-csi-node-rppvz                     Container image "k8s.gcr.io/sig-storage/csi-node-driver-registrar:v2.1.0" already present on machine
> 45m         Normal    Created            pod/freenas-nfs-democratic-csi-node-rppvz                     Created container driver-registrar
> 45m         Normal    Started            pod/freenas-nfs-democratic-csi-node-rppvz                     Started container driver-registrar
> 45m         Normal    Pulled             pod/freenas-nfs-democratic-csi-node-rppvz                     Container image "docker.io/busybox:1.32.0" already present on machine
> 45m         Normal    Created            pod/freenas-nfs-democratic-csi-node-rppvz                     Created container cleanup
> 45m         Normal    Started            pod/freenas-nfs-democratic-csi-node-rppvz                     Started container cleanup
> 42m         Normal    BackOff            pod/freenas-nfs-democratic-csi-node-rppvz                     Back-off pulling image "democraticcsi/democratic-csi:latest"
> 114s        Warning   Failed             pod/freenas-nfs-democratic-csi-node-rppvz                     Error: ImagePullBackOff
> 42m         Warning   Failed             pod/freenas-nfs-democratic-csi-node-rppvz                     Failed to pull image "democraticcsi/democratic-csi:latest": rpc error: code = Unknown desc = Error reading manifest latest in docker.io/democraticcsi/democratic-csi: toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit
> 50m         Normal    LeaderElection     lease/org-democratic-csi-iscsi                                zfs-iscsi-democratic-csi-controller-64fc47b6dd-8t9sl became leader
> 50m         Normal    LeaderElection     lease/org-democratic-csi-nfs                                  freenas-nfs-democratic-csi-controller-697cb8cd9b-886h7 became leader
> 50m         Normal    Scheduled          pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-8t9sl      Successfully assigned democratic-csi/zfs-iscsi-democratic-csi-controller-64fc47b6dd-8t9sl to node-b1.okd4.rdu.bretm.net
> 50m         Normal    AddedInterface     pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-8t9sl      Add eth0 [10.128.3.153/23]
> 50m         Normal    Pulled             pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-8t9sl      Container image "k8s.gcr.io/sig-storage/csi-provisioner:v2.1.0" already present on machine
> 50m         Normal    Created            pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-8t9sl      Created container external-provisioner
> 50m         Normal    Started            pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-8t9sl      Started container external-provisioner
> 50m         Normal    Pulled             pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-8t9sl      Container image "k8s.gcr.io/sig-storage/csi-resizer:v1.1.0" already present on machine
> 50m         Normal    Created            pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-8t9sl      Created container external-resizer
> 50m         Normal    Started            pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-8t9sl      Started container external-resizer
> 50m         Normal    Pulled             pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-8t9sl      Container image "k8s.gcr.io/sig-storage/csi-snapshotter:v3.0.3" already present on machine
> 50m         Normal    Created            pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-8t9sl      Created container external-snapshotter
> 50m         Normal    Started            pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-8t9sl      Started container external-snapshotter
> 50m         Normal    Pulling            pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-8t9sl      Pulling image "democraticcsi/democratic-csi:latest"
> 50m         Normal    Pulled             pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-8t9sl      Successfully pulled image "democraticcsi/democratic-csi:latest" in 849.77569ms
> 50m         Normal    Created            pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-8t9sl      Created container csi-driver
> 50m         Normal    Started            pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-8t9sl      Started container csi-driver
> 46m         Normal    Killing            pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-8t9sl      Stopping container external-provisioner
> 46m         Normal    Killing            pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-8t9sl      Stopping container csi-driver
> 46m         Normal    Killing            pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-8t9sl      Stopping container external-snapshotter
> 46m         Normal    Killing            pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-8t9sl      Stopping container external-resizer
> 50m         Normal    Killing            pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-dh7zw      Stopping container external-provisioner
> 50m         Normal    Killing            pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-dh7zw      Stopping container csi-driver
> 50m         Normal    Killing            pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-dh7zw      Stopping container external-snapshotter
> 50m         Normal    Killing            pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-dh7zw      Stopping container external-resizer
> 46m         Normal    Scheduled          pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-mv8rr      Successfully assigned democratic-csi/zfs-iscsi-democratic-csi-controller-64fc47b6dd-mv8rr to node-a1.okd4.rdu.bretm.net
> 46m         Normal    AddedInterface     pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-mv8rr      Add eth0 [10.131.0.14/23]
> 46m         Normal    Pulled             pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-mv8rr      Container image "k8s.gcr.io/sig-storage/csi-provisioner:v2.1.0" already present on machine
> 46m         Normal    Created            pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-mv8rr      Created container external-provisioner
> 46m         Normal    Started            pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-mv8rr      Started container external-provisioner
> 46m         Normal    Pulled             pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-mv8rr      Container image "k8s.gcr.io/sig-storage/csi-resizer:v1.1.0" already present on machine
> 46m         Normal    Created            pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-mv8rr      Created container external-resizer
> 46m         Normal    Started            pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-mv8rr      Started container external-resizer
> 46m         Normal    Pulled             pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-mv8rr      Container image "k8s.gcr.io/sig-storage/csi-snapshotter:v3.0.3" already present on machine
> 46m         Normal    Created            pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-mv8rr      Created container external-snapshotter
> 46m         Normal    Started            pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-mv8rr      Started container external-snapshotter
> 43m         Normal    Pulling            pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-mv8rr      Pulling image "democraticcsi/democratic-csi:latest"
> 15m         Warning   Failed             pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-mv8rr      Failed to pull image "democraticcsi/democratic-csi:latest": rpc error: code = Unknown desc = Error reading manifest latest in docker.io/democraticcsi/democratic-csi: toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit
> 41m         Warning   Failed             pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-mv8rr      Error: ErrImagePull
> 93s         Normal    BackOff            pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-mv8rr      Back-off pulling image "democraticcsi/democratic-csi:latest"
> 6m26s       Warning   Failed             pod/zfs-iscsi-democratic-csi-controller-64fc47b6dd-mv8rr      Error: ImagePullBackOff
> 50m         Normal    SuccessfulCreate   replicaset/zfs-iscsi-democratic-csi-controller-64fc47b6dd     Created pod: zfs-iscsi-democratic-csi-controller-64fc47b6dd-8t9sl
> 46m         Normal    SuccessfulCreate   replicaset/zfs-iscsi-democratic-csi-controller-64fc47b6dd     Created pod: zfs-iscsi-democratic-csi-controller-64fc47b6dd-mv8rr
> 42m         Warning   NodeNotReady       pod/zfs-iscsi-democratic-csi-node-vw2bm                       Node is not ready
> 38m         Normal    Pulling            pod/zfs-iscsi-democratic-csi-node-vw2bm                       Pulling image "democraticcsi/democratic-csi:latest"
> 37m         Warning   Failed             pod/zfs-iscsi-democratic-csi-node-vw2bm                       Failed to pull image "democraticcsi/democratic-csi:latest": rpc error: code = Unknown desc = Error reading manifest latest in docker.io/democraticcsi/democratic-csi: toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit
> 36m         Warning   Failed             pod/zfs-iscsi-democratic-csi-node-vw2bm                       Error: ErrImagePull
> 40m         Normal    Pulled             pod/zfs-iscsi-democratic-csi-node-vw2bm                       Container image "k8s.gcr.io/sig-storage/csi-node-driver-registrar:v2.1.0" already present on machine
> 40m         Normal    Created            pod/zfs-iscsi-democratic-csi-node-vw2bm                       Created container driver-registrar
> 40m         Normal    Started            pod/zfs-iscsi-democratic-csi-node-vw2bm                       Started container driver-registrar
> 40m         Normal    Pulled             pod/zfs-iscsi-democratic-csi-node-vw2bm                       Container image "docker.io/busybox:1.32.0" already present on machine
> 40m         Normal    Created            pod/zfs-iscsi-democratic-csi-node-vw2bm                       Created container cleanup
> 40m         Normal    Started            pod/zfs-iscsi-democratic-csi-node-vw2bm                       Started container cleanup
> 26m         Normal    BackOff            pod/zfs-iscsi-democratic-csi-node-vw2bm                       Back-off pulling image "democraticcsi/democratic-csi:latest"
> 37m         Warning   Failed             pod/zfs-iscsi-democratic-csi-node-vw2bm                       Error: ImagePullBackOff
> 48m         Warning   NodeNotReady       pod/zfs-iscsi-democratic-csi-node-ztxhx                       Node is not ready
> 44m         Normal    Pulling            pod/zfs-iscsi-democratic-csi-node-ztxhx                       Pulling image "democraticcsi/democratic-csi:latest"
> 43m         Warning   Failed             pod/zfs-iscsi-democratic-csi-node-ztxhx                       Failed to pull image "democraticcsi/democratic-csi:latest": rpc error: code = Unknown desc = Error reading manifest latest in docker.io/democraticcsi/democratic-csi: toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit
> 16m         Warning   Failed             pod/zfs-iscsi-democratic-csi-node-ztxhx                       Error: ErrImagePull
> 45m         Normal    Pulled             pod/zfs-iscsi-democratic-csi-node-ztxhx                       Container image "k8s.gcr.io/sig-storage/csi-node-driver-registrar:v2.1.0" already present on machine
> 45m         Normal    Created            pod/zfs-iscsi-democratic-csi-node-ztxhx                       Created container driver-registrar
> 45m         Normal    Started            pod/zfs-iscsi-democratic-csi-node-ztxhx                       Started container driver-registrar
> 45m         Normal    Pulled             pod/zfs-iscsi-democratic-csi-node-ztxhx                       Container image "docker.io/busybox:1.32.0" already present on machine
> 45m         Normal    Created            pod/zfs-iscsi-democratic-csi-node-ztxhx                       Created container cleanup
> 45m         Normal    Started            pod/zfs-iscsi-democratic-csi-node-ztxhx                       Started container cleanup
> 42m         Normal    BackOff            pod/zfs-iscsi-democratic-csi-node-ztxhx                       Back-off pulling image "democraticcsi/democratic-csi:latest"
> 6m56s       Warning   Failed             pod/zfs-iscsi-democratic-csi-node-ztxhx                       Error: ImagePullBackOff
> [root@okd4-services ~]#

**Comment by travisghansen** [comment]
So I'm guessing the error messages are a result of the 3rd pod down not being in a fully running state.

These are the relevant sockets on a standard `/var/lib/kubelet` data dir install:

```
[root@node01 ~]# ls -l /var/lib/kubelet/plugins_registry/*
srwx------ 1 root root 0 May 22 22:04 /var/lib/kubelet/plugins_registry/org.democratic-csi.nfs-reg.sock
[root@node01 ~]# ls -l /var/lib/kubelet/plugins/org.democratic-csi.nfs/
srwxr-xr-x 1 root root 0 May 22 22:04 csi.sock
```

I know it doesn't directly resolve the issue, but are you running an image cache instance on your network? It's super easy to setup and will likely help tremendously in mitigating the risk. I've got basic instructions if you'd like them.

**Comment by travisghansen** [comment]
Images are now available on ghcr as well starting with `v1.3.0`.

**Comment by Routhinator** [comment]
What are the image paths? The chart should likely default to ghcr
